### PR TITLE
Adding support for gradient runs

### PR DESCRIPTION
**Pull request by jmorrisnrel**
Adding support to Engage to build and run multiple-year gradient runs with each subsequent year adding onto the previous year's results.

A run for each year is built at the start of the process, and the celery/batch run task for the next year is kicked off at the end of each previous run.

Added new run group field to help group and manage the multiple run objects in a gradient run

### CODE REVIEW

File: calliope_app/api/views/outputs.py [hunk 1]
```diff
@@ -267,6 +290,29 @@ def optimize(request):
             run.status = task_status.QUEUED
             run.save()
             payload = {"task_id": response.get("jobId")}
+            if run.group != '':
+                future_runs = Run.objects.filter(group=run.group,year__gt=run.year).order_by('year')
+                for next_run in future_runs:
+                    if next_run.status == task_status.BUILT:
```
Comment (jgu2) [inline calliope_app/api/views/outputs.py:296]: nit: Reduce deep indent, can use early returns. e.g.,
```py
if not run.group:
    break

futures_runs = ...
for next_run in future_runs:
    if next_run.status != task_status.BUILT:
        break
    logger.info(...)
```

File: calliope_app/api/tasks.py [hunk 2]
```diff
@@ -576,6 +576,36 @@ def run_model(run_id, model_path, user_id, *args, **kwargs):
...
+        for next_run in future_runs:
+            if next_run.status == task_status.BUILT:
+                logger.info("Found a subsequent gradient model for year %s.",next_run.year)
+                apply_gradient(run.inputs_path,save_outputs,next_run.inputs_path,run.year,next_run.year,logger)
+                if next_run == future_runs.first():
+                    model_path = os.path.join(next_run.inputs_path, "model.yaml")
+                    environment = next_run.compute_environment
+                    logger.info("Model run %s is ready to run.",next_run.id)
+
+                    if environment.type == "Celery Worker":
```
Comment (jgu2) [inline calliope_app/api/tasks.py:591]: nit: reduce deep indents using early returns, see example below.

File: calliope_app/api/calliope_utils.py [hunk 3]
```diff
@@ -431,14 +435,171 @@ def _yaml_outputs(model_path, outputs_dir):
...
+                            new_energy_cap_max = new_loc_tech['constraints'].get('energy_cap_max',new_tech['constraints'].get('energy_cap_max',0))
+                            new_storage_cap_min = new_loc_tech['constraints'].get('storage_cap_min',new_tech['constraints'].get('storage_cap_min',0))
+                            new_storage_cap_max = new_loc_tech['constraints'].get('storage_cap_max',new_tech['constraints'].get('storage_cap_max',0))
+                        else:
+                            new_energy_cap_min = new_tech['constraints'].get('energy_cap_min',0)
+                            new_energy_cap_max = new_tech['constraints'].get('energy_cap_max',0)
+                            new_storage_cap_min = new_tech['constraints'].get('storage_cap_min',0)
+                            new_storage_cap_max = new_tech['constraints'].get('storage_cap_max',0)
+
+                        if new_loc_tech == None:
```
Comment (jgu2) [inline calliope_app/api/calliope_utils.py:504]:
```py
if new_loc_tech is None:
    ...
```

File: calliope_app/api/calliope_utils.py [hunk 4]
```diff
@@ -431,14 +435,171 @@ def _yaml_outputs(model_path, outputs_dir):
...
-                            model[tl][l]['techs'][t]['results']['energy_cap'] = float(r_df.loc[(r_df['locs'] == l) &
-                                                                                (r_df['techs'] == t)]['energy_cap'].values[0])
+                            model[tl][l]['techs'][t]['results'][v+'_equals'] = float(r_df.loc[(r_df['locs'] == l) &
+                                                                                (r_df['techs'] == t)][v].values[0])
     if has_outputs:
-        yaml.dump(model, open(os.path.join(outputs_dir,'model_results.yaml'),'w+'), default_flow_style=False)
\ No newline at end of file
+        yaml.dump(model, open(os.path.join(outputs_dir,'model_results.yaml'),'w+'), default_flow_style=False)
+
+def apply_gradient(old_inputs,old_results,new_inputs,old_year,new_year,logger):
```
Comment (jgu2) [inline calliope_app/api/calliope_utils.py:452]: This function is long and complicated with deep nested `for ... if ... else`, computational complexity is O(n^4), easy to be buggy. Normally when complexity is reaching to O(n^3), then consider to refactor, normally use a map or hash to query and reduce a level of `for` loop.

Comment (jmorrisnrel) [reply calliope_app/api/calliope_utils.py:452]: I think the complexity is better than O(n^4) as it's basically looping through all of the parameters in a very specific dictionary/YAML file format which is what results in all of the `for` loops and `if ... else` conditionals. The for loops are looping through the sublevels of the dict/YAML for each element of the dict so it doesn't technically revisit any of the levels.

It's definitely confusing and not very readable but I'll need to work out the best way to implement a map that doesn't just make it harder to understand what's going on. I've tested this code pretty heavily on the HPC at this point and feel confident that it works as intended enough to get the code deployed on dev to test the Batch code

File: calliope_app/api/models/outputs.py [hunk 5]
```diff
@@ -54,6 +54,7 @@ class Meta:
         null=True,
         on_delete=models.SET_NULL
     )
+    group = models.TextField(blank=True)
```
[the remark this answers is not in the document]
Comment (jgu2) [reply calliope_app/api/models/outputs.py:57]: Would `group` to be nullable? if not, what's the default value would be for db migration with existing records.

[the remark this answers is not in the document]
Comment (jmorrisnrel) [reply calliope_app/api/models/outputs.py]: I default it to blank on new runs but I've made it nullable for older runs which shouldn't affect anything else as it's only used on run creation.

File: calliope_app/api/migrations/0058_alter_run_group.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+        ('api', '0057_run_group'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='run',
+            name='group',
+            field=models.TextField(blank=True, null=True),
+        ),
+    ]
```
Comment (jgu2) [inline calliope_app/api/migrations/0058_alter_run_group.py:18]: nit: For final PR with migrations, normally create one migration file, that is
1. delete `0057` and `0058`.
2. run `python manage makemigrations` to create a new `0057` with `group` added and `blank=True, null=True`.
3. run `python manage migrate --fake` to do a fake migration the new `0057`, as your local db already was already migrated with old `0057` and `0058`.

The purpose is to reduce the number of db migrations, and reduce the risk of migration issue, as dealing with db migration issue is painful. It's a best practice. Well, it's not a big deal this time, as we already have a long list of migrations files, it needs to squash the migrations, not with PR, maybe in a future task.

**Review by jgu2 — APPROVED** [review]

**Pull request merged**